app/controllers/sessions_controller.rb

```diff
@@ -1,21 +1,92 @@
+require 'open-uri'
 class SessionsController < ApplicationController
   def new
     @page_title = 'Sign In | Haxxor News'
   end
 
   def create
-    user = User.find_by(username: params[:username])
+    @oauth = OAuthUser.where(username: params[:username]).first 
+    unless @oauth
+      user = User.find_by(username: params[:username])
+      if user && user.authenticate(params[:password])
+        session[:user_id] = user.id
+        redirect_to articles_path
+      else
+        flash[:alert] = "Invalid username or password."
+        redirect_to new_sessions_path
+      end
+    else
+      flash[:alert] = "#{@oauth.username}, please sign in using #{@oauth.provider.capitalize}"
+      redirect_to new_sessions_path
+    end
+  end
 
-    if user && user.authenticate(params[:password])
-      session[:user_id] = user.id
-      redirect_to articles_path
+  def redirect
+    @response = auth_hash
+    info = @response["info"]
+    #Checking whether the user from provider has already been created thus bypassing everything
+    @oauth = OAuthUser.where(email: info["email"]).first
+    
+    unless @oauth
+      if @response["provider"] == "facebook"
+        user_params = {
+          username: create_username(info["name"]),
+          email: info["email"]
+          }
+        @profile_img = grab_image(info["image"])
+      end
+      
+      #Checking whether or not the user already exists based on given parameters from provider
+      @user_with_name = User.where(username: user_params[:username]).first
+      @user_with_email = User.where(email: user_params[:email]).first
+      unless @user_with_name || @user_with_email
+        @user = User.new(user_params)
+        @user.oauth_creation = true
+        if @profile_img
+          @user.profile_img.attach(io: @profile_img, filename: "#{user_params[:username]}profileimg.jpeg")
+        end
+                
+        if @user.save
+          OAuthUser.create(username: @user.username, email: @user.email, provider: @response["provider"])
+          session[:user_id] = @user.id
+          @user.oauth_creation = false
+          redirect_to articles_path
+        end
+      else
+        flash[:alert] = "User already exists. Please sign in natively."
+        redirect_to new_sessions_path
+      end
     else
-      render 'new'
+      session[:user_id] = User.where(email: @oauth.email).first.id
+      flash[:notice] = "Signed in!"
+      redirect_to articles_path
     end
   end
```

Comment:
This whole thing is pretty long and tough to make sense of. I'd recommend using `if` instead of `unless` as that tends to be easier to reason over (for me at least). It also would be nice to break this big method into a bunch of smaller ones, possibly even a class to handle some of the lookups and database updates.

Reply:
Moved a lot of the lookups which made up a lot of the method into classes. Also the OAuthUser:User relation condensed a lot